script/update_snapshots

```diff
@@ -7,7 +7,8 @@ config_file=$app_root/config/mysql-$RAILS_ENV.cnf
 public=$app_root/public
 
 echo "image_id	observation_id" > $public/observation_images.csv
-mysql --defaults-extra-file=$config_file -q -s -e 'select image_id, observation_id from observation_images' >> $public/images_observations.csv
+mysql --defaults-extra-file=$config_file -q -s -e 'select image_id, observation_id from observation_images' >> $public/observation_images.csv
+cp $public/observation_images.csv $public/images_observations.csv
```

Comment:
Huh???  Does "echo > file" not clobber the file???  Oh!!!  Nimmo must have changed one of the "images_observations" to "observation_images" but not both.  Can we simply change that back instead of doing the copy?  (Shouldn't it have been a mv -f instead of a cp, anyway?  Won't this code just keep observation_images.csv around and grow it ad infinitum?

I would like to see this file remain images_observations.csv.  The fact that we changed the table name internally need not affect the various users that use this file.

Summary: Please revert line 10, remove the copy, and fix line 9 to output the header to the correct file, images_observations.csv.

Reply:
Okay, I've finally gotten my head fully around this.  (I think!)

The original bug was introduced when Nimmo changed one of the filenames from images_observations.csv to observation_images.csv, but not the other.  Nice and simple.  It's the particular fix that had me so confused.

Maybe you are deliberately "keeping the bug" because it creates a new file (observation_images.csv) separately that you can then move into place atomically?  If that is the reason wouldn't it be better to move than copy?  Does keeping the copy around serve some purpose I've not thought of?  And shouldn't we also do it for all the other files, not just images_observations.csv?  If that one file can benefit from being changed atomically, wouldn't they all?  In fact, maybe we should create the whole *suite* of files separately first, then move them all into place in one go.  They kind of need to be all the same date.  For that matter, if we're really being strict, wouldn't we want to lock the database to prevent any new writes while we dump all the tables?  How careful do we need to be?  And lastly, wouldn't the intent be clearer if we called the new file "images_observations.csv.new" instead of "observation_images.csv"?  That's such a random and/or potentially confusing name in this context!

Reply:
So my thinking was let's first create observation_images.csv correctly (since it appears to be the shiny new name), but then maintain backwards compatibility by immediately copying it to the old name.  This would also clean up both files that are currently getting created incorrectly.  If we're at all worried about the disk space or the time it takes to do the copy, then we could use a hard (or a symbolic) link instead of a making a copy.  However, at the end of the day I don't care as long as it doesn't just keep growing.

I was definitely not thinking about atomicity and I don't think we need to.  Looks like the worst possible case would be an observation getting created after the join table gets dumped which doesn't seem like a bit deal since any associated images would just appear in the next dump.

Reply:
I approve of the superior analysis going on here. Sorry to introduce the bug, i don't think I would have intentionally changed a filename output like that, but i should have caught it.

I do like the idea of having it available under the new name of the table, and the old one for compatibility. Who uses these csv's, by the way?